llvm/lib/IR/ConstantRange.cpp

```diff
@@ -145,6 +145,14 @@ ConstantRange ConstantRange::makeAllowedICmpRegion(CmpInst::Predicate Pred,
   }
 }
 
+ConstantRange ConstantRange::makeAsymmetricICmpRegion(CmpPredicate Pred,
+                                                      const ConstantRange &CR) {
+  if (Pred.hasSameSign() && ICmpInst::isRelational(Pred))
+    return makeAllowedICmpRegion(Pred, CR).unionWith(makeAllowedICmpRegion(
```

Comment:
ICmp with `samesign` should give a tighter set.
Example: `makeAllowedICmpRegion` returns `[0, 4)` for `icmp samesign ult X, [2, 5)`, but `makeAsymmetricICmpRegion` returns `[INT_MIN, 4)`.

Reply:
Should use #126528, I think.

Reply:
Scratch that. After thinking some more, I think I'm missing an inverse on the sign-flipped check.

Reply:
I think `icmp samesign ult X, [2, 5)` should return the disjoint union of [0, 4) and [INT_MIN, -123). We have no way to represent a disjoint union as a ConstantRange, so I think we need a pair of ConstantRanges to check against? Not 100% sure, so would appreciate a sanity-check.